File: c#/src/Sodium/Sodium/MutableListener.cs
namespace Sodium
{
    public class MutableListener : IListener
    {
        private readonly WeakMutableListener weakMutableListener = new WeakMutableListener();
        private IListener listener;

        public void SetListener(IListener listener)
        {
            this.listener = listener;
            this.weakMutableListener.WeakListener = listener.GetListenerWithWeakReference();
        }

        public void Unlisten()
        {
            this.listener?.Unlisten();
        }

        public IListenerWithWeakReference GetListenerWithWeakReference() => this.weakMutableListener;

        private class WeakMutableListener : IListenerWithWeakReference
        {
            public IListenerWithWeakReference WeakListener;

            public void Unlisten()
            {
                this.WeakListener?.Unlisten();
            }
        }
    }
}

File: f#/src/SodiumFRP.FSharp/_Root/MutableListener.fs
namespace SodiumFRP.FSharp

type internal WeakMutableListener =
    val mutable public WeakListener : IListenerWithWeakReference option
    new() = { WeakListener = None }
    interface IListenerWithWeakReference with
        member this.Unlisten() =
            this.WeakListener |> Option.iter Listener.unlistenWeak

type MutableListener() =
    let weakMutableListener = WeakMutableListener()
    let mutable listener = None
    member internal __.SetListener l =
        listener <- Some l
        weakMutableListener.WeakListener <- Some <| Listener.getListenerWithWeakReference l
    member internal __.ClearListener () =
        listener <- None
        weakMutableListener.WeakListener <- None
    interface IListener with
        member __.Unlisten() =
            listener |> Option.iter Listener.unlisten
        member __.GetListenerWithWeakReference () = upcast weakMutableListener

module MutableListener =
    let setListener l (m : MutableListener) = m.SetListener l
    let clearListener (m : MutableListener) = m.ClearListener ()
    let unlisten (m : MutableListener) = Listener.unlisten m
    let getListenerWithWeakReference (m : MutableListener) = Listener.getListenerWithWeakReference m
